### ISSUE TITLE
Multiple Rules Got Merged into One

### ISSUE THREAD

**Issue by binotaliu** [tailwindlabs/tailwindcss#3964]
### What version of @tailwindcss/jit are you using?

v0.1.18

### What version of Node.js are you using?

v14.16.0

### What build tool (or framework if it abstracts the build tool) are you using?

laravel-mix v6.0.13

### What browser are you using?

N/A

### What operating system are you using?

macOS

### Reproduction repository

https://github.com/binotaliu/tailwindcss-test

given following source ([test.css](https://github.com/binotaliu/tailwindcss-test/blob/4cc053827dfafb08e4e1809e5691d44262f30518/test.css)) :

```css
.content {
  ul:not(:first-of-type),
  .content-list ul {
    @apply list-none m-0 mb-[2.5em] p-0;
    li {
      @apply relative block pl-5 mb-4 text-[.89em] text-gray-700 leading-[1.714em];
      a {
        @apply no-underline;
      }
      code {
        @apply text-[.875em];
      }
    }
  }
}
```

with `NODE_ENV=production`, the compiled result is ([dist.css](https://github.com/binotaliu/tailwindcss-test/blob/4cc053827dfafb08e4e1809e5691d44262f30518/dist.css)):
```css
.content .content-list ul,
.content ul:not(:first-of-type) {
  margin: 0 0 1rem;
  list-style-type: none;
  padding: 0;
  padding: 0 0 0 1.25rem;
  display: block;
  font-size: 0.89em;
  line-height: 1.714em;
  --tw-text-opacity: 1;
  color: rgba(55, 65, 81, var(--tw-text-opacity));
  text-decoration: none;
  font-size: 0.875em
}
```

----

I tried to not use `postcss-nested`, so use following source ([test-without-nested.css](https://github.com/binotaliu/tailwindcss-test/blob/4cc053827dfafb08e4e1809e5691d44262f30518/test-without-nested.css)):
```css
@tailwind components;
@tailwind utilities;
@layer components {
  .content .content-list ul,
  .content ul:not(:first-of-type) {
    @apply list-none m-0 mb-[2.5em] p-0;
  }
  
.content .content-list ul li,
  .content ul:not(:first-of-type) li {
    @apply relative block pl-5 mb-4 text-[.89em] text-gray-700 leading-[1.714em];
  }
  
.content .content-list ul li a,
  .content ul:not(:first-of-type) li a {
    @apply no-underline}
  
.content .content-list ul li code,
  .content ul:not(:first-of-type) li code {
    @apply text-[.875em]}
}
```

this looks fine, and I can compile it with `@tailwindcss/jit` and get a expected result ([dist-without-nested.css](https://github.com/binotaliu/tailwindcss-test/blob/4cc053827dfafb08e4e1809e5691d44262f30518/dist-without-nested.css)) :
```css
.content .content-list ul,
.content ul:not(:first-of-type) {
  margin: 0 0 2.5em;
  list-style-type: none;
  padding: 0
}

.content .content-list ul li,
.content ul:not(:first-of-type) li {
  position: relative;
  margin-bottom: 1rem;
  display: block;
  padding-left: 1.25rem;
  font-size: 0.89em;
  line-height: 1.714em;
  --tw-text-opacity: 1;
  color: rgba(55, 65, 81, var(--tw-text-opacity))
}

.content .content-list ul li a,
.content ul:not(:first-of-type) li a {
  text-decoration: none
}

.content .content-list ul li code,
.content ul:not(:first-of-type) li code {
  font-size: 0.875em
}
```

----

Looks like it is a bug that only happen when using in combination with nested rules and `,` selector.

**Comment by therealgilles** [comment]
I'm having the same issue.

**Comment by therealgilles** [comment]
From my analysis, the problem lies with Tailwindcss, not with postcss-nested.

**Comment by therealgilles** [comment]
Here is my css:

```CSS
	.dataTables_wrapper .dataTables_filter label,
	.dataTables_length label {
		@apply text-base;
		line-height: 40px;

		input[type='search'] {
			@apply bg-white;
		}
	}
```

If I disable Tailwindcss, I get this, which looks correct:

```CSS
.analytics .dataTables_length label,.analytics .dataTables_wrapper .dataTables_filter label{@apply text-base;line-height:40px;}
.analytics .dataTables_length label input[type=search],.analytics .dataTables_wrapper .dataTables_filter label input[type=search]{@apply bg-white}
```

When I enable Tailwindcss, I get this, which is incorrect:

```CSS
.analytics .dataTables_length label,.analytics .dataTables_wrapper .dataTables_filter label{font-size:16px;line-height:24px;line-height:40px;--tw-bg-opacity:1;background-color:rgba(255,255,255,var(--tw-bg-opacity))}
```

The input[type=search] white background property got merged with the label parent properties.

**Comment by therealgilles** [comment]
It would be great if someone from the tailwindcss team could look into this. From my experience, this problem exists even without enabling JIT mode. It's quite frustrating as it is impossible to figure what will work and what won't. I tried to switch post-nested with post-sass and got the same incorrect result.

**Comment by adamwathan** [comment]
Planning to look at it when I have time, in the mean time though I would just recommend writing your CSS to not use nesting and the problem will definitely go away. Nesting isn't really a standard feature of CSS so Tailwind doesn't fully "understand" it especially with `@apply`, and the nesting plugins all have different bugs that make them behave weirdly when used in front of Tailwind to try and flatten the CSS before Tailwind sees it.

**Comment by adamwathan** [comment]
Yeah it used to work better before the new event system, IMO that approach is flawed for transformations that are super structural like nesting. Maybe tomorrow I will fork one of the nesting plugins and fix the bugs and we can maintain our own or something.

**Comment by therealgilles** [comment]
@adamwathan: Did you find some time to look into this?

**Comment by adamwathan** [comment]
I did, we will have a public solution for this in v2.2 but in the mean time you can install our canary release:

`npm install -D tailwindcss@canary`

And then importantly, make sure you don't use `postcss-nested` directly but instead use this little custom plugin (we are gonna wrap this up as `@tailwindcss/postcss-nested`)

```js
// postcss.config.js
const postcss = require("postcss");
const nested = require("postcss-nested");

module.exports = {
  plugins: [
    function (root) {
      root.walkAtRules("screen", (rule) => {
        rule.name = "media";
        rule.params = `screen(${rule.params})`;
      });

      return postcss([postcssNested(opts)]).process(root, result.opts);
    },
    require('tailwindcss'),
    require('autoprefixer'),
  ],
};
```

**Comment by therealgilles** [comment]
@adamwathan: this is honestly the best news! Could you let us know what you found out?
I may wait for v2.2 to come out. Oh and is there a similar issue with postcss-nested-ancestors or postcss-nesting?

**Comment by adamwathan** [comment]
So the main issue is just that `postcss-nested` moved to the new PostCSS visitor API, which causes it to run in a non-deterministic order. That wrapper plugin I showed above will force it to run in the right order, guaranteeing that all of the CSS has been "un-nested" by the time Tailwind sees it.

The other piece is that nesting plugins won't automatically treat `@screen` rules the same as media queries, and won't "bubble" them out by default. So we've added a new `screen()` function in Tailwind that let's you do this:

```css
@media screen(sm) {
  /* ... */
}
```

...instead of using `@screen`. Because it's a real media query, nesting plugins will bubble it out properly.

The wrapper plugin I showed above will automatically convert any `@screen` rules into `@media screen(...)` rules instead, before passing that CSS into the nesting plugin. Then the nesting plugin will unnest everything, and Tailwind will process the unnested CSS perfectly, converting all `screen()` calls into the proper media queries for you.

We're going to bundle that wrapper plugin up ourself as `@tailwindcss/postcss-nested` and recommend people use that instead.

I haven't tried the other plugins but generally, using `@media screen()` instead of `@screen` + using our canary build so that `screen()` is actually supported should make it possible to use any existing PostCSS plugin that handles that sort of thing.

Going to close since this is resolved in master 👍🏻
